### ISSUE TITLE
B006 (MutableArgumentDefault) should exempt parameters with immutable type annotations

### ISSUE THREAD

**Issue by andersk** [charliermarsh/ruff#572]
B006 aims to prevent bugs that arise from mutating the default value of a function’s parameter, with unintended effects on all future calls to that function. But for a parameter that’s annotated with an immutable abstract type such as `Sequence[int]`, mypy already prevents the value from being mutated. So such parameters should be exempt from B006. For example:

```python
from typing import Sequence

def f(a: Sequence[int] = [1, 2, 3]):  # should not raise B006
    for x in a:
        print(x)
```

This was suggested at https://github.com/PyCQA/flake8-bugbear/issues/137#issuecomment-805724850, but [wasn’t implemented](https://github.com/PyCQA/flake8-bugbear/issues/137#issuecomment-808161372) due to flake8-bugbear lacking the machinery to accurately track `typing` imports. Since ruff has that now (#533), perhaps this is easy to implement in ruff.

Types that should be considered immutable:

```python
object
collections.abc.Container[T]
collections.abc.Iterable[T]
collections.abc.Reversible[T]
collections.abc.Sized
collections.abc.Collection[T]
collections.abc.Sequence[T]
collections.abc.Set[T]
collections.abc.Mapping[K, V]
typing.Container[T]
typing.Iterable[T]
typing.Reversible[T]
typing.Sized
typing.Collection[T]
typing.Sequence[T]
typing.AbstractSet[T]  # not typing.Set[T]
typing.Mapping[K, V]
typing.Optional[I]  # where I is any of the above
I | None  # where I is any of the above
typing.Annotated[I, …]  # where I is any of the above
```

**Comment by charliermarsh** [comment]
(Agree.)

**Comment by harupy** [comment]
> mypy already prevents the value from being mutated

There might be edge cases where mypy doesn't prevent the value from being mutated.

```python
# a.py

from typing import Sequence


def untyped_func_imported_from_third_party_library(a):
    a.append(1)


def f(a: Sequence[int] = [1, 2, 3]):
    untyped_func_imported_from_third_party_library(a)
    print(a)


f()
f()
```


```
> mypy --version
mypy 0.982 (compiled: yes)

> mypy a.py
Success: no issues found in 1 source file

> python a.py
[1, 2, 3, 1]
[1, 2, 3, 1, 1]
```

**Comment by andersk** [comment]
Sure, Python’s gonna Python, so mypy will always be an incomplete solution (although there are options like `--strict` that make it less incomplete). But what I really meant was that, by annotating `a: Sequence[int]`, the programmer has at least declared an *intention* to avoid mutating `a`, and it’s then up to the programmer to enforce this intention, via an appropriate mypy configuration or otherwise.
